File: javascript/LCA.js
//implemented based off of BST class made in Second Year

/**
 * @class Node
 *      a class used to store the data for different nodes of the binary tree.
 */

class Node {
    constructor(key) {
        this.left = null;
        this.right = null;
        this.key = key;
    }
}

let root = null;
let N = 0;

/**
 *  Insert key into BinaryTree.
 *  If key already exists, update with new value.
 *   
 *  @param key the key to insert
 */

function put(key) {
    root = put(root, key);
    N++;
}

function put(x, key) {
    if(x == null) return new Node(key);
    if      (x.key > key) x.left = put(x.left, key);
    else if (x.key < key) x.right = put(x.right, key);
    return x;
}


/**
 *  Search BST for given key.
 *  What is the value associated with given key?
 *
 *  @param key the search key
 *  @return value associated with the given key if found, or null if no such key exists.
 */
function get(key) {
    return get(root, key);
}

function get(x, key) {
    if(x == null) return false;
    if      (x.key > key) return get(x.left, key);
    else if (x.key < key) return get(x.right, key);
    return true;
}

/**
 * finds the Lowest Common Ancestor of two given key values
 * @param k1 the key value of the first LCA node to be found
 * @param k2 the key value of the second LCA node to be found
 * @return the node that is the LCA of k1 and k2
 */

function findLCA(k1, k2) {
    if(get(k1) && get(k2)) {
        return findLCA(root, k1, k2);
    }
    else return null;
}

function findLCA(x, k1, k2) {
    if(x == null) return null;
    if(x.key == k1 || x.key == k2) return x;
    leftLCA = findLCA(x.left, k1, k2);
    rightLCA = findLCA(x.right, k1, k2);
    if(leftLCA != null && rightLCA != null) return x;
    if(leftLCA != null) return leftLCA;
    else return rightLCA;
}

function Main() {
    put(4);
    put(2);
    put(3);
    put(1);
    put(6);
    put(5);
    put(7);
    lca = findLCA(3,7);
    if(lca != null) {
        document.write("LCA of key:3 and key:7 = " + lca.key);
    }
    else {
        document.write("LCA of key:3 and key:7 does not exist");
    }
    lca = findLCA(4,10);
    if(lca != null) {
        document.write("LCA of key:4 and key:10 = " + lca.key);
    }
    else {
        document.write("LCA of key:4 and key:10 does not exist");
    }
}

File: Python/LCA.py
class Node:
    def __init__(self,key):
        self.key = key
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self,key):
        self.root = Node(key)

    def put(self, root, key):
        if root is None: 
            return Node(key)
        if root.key > key: 
            root.left = self.put(root.left, key)
        elif root.key < key: 
            root.right = self.put(root.right, key)
        return root  
    
    def get(self, root, key):
        if root is None: 
            return False
        if root.key > key: 
            return self.get(root.left, key)
        elif root.key < key: 
            return self.get(root.right, key)
        return True  

    def findLCA(self,root, k1, k2):
        if root is None: 
            return None
        if root.key is k1 or root.key is k2: 
            return root
        leftLCA = self.findLCA(root.left,k1,k2)
        rightLCA = self.findLCA(root.right,k1,k2)
        if leftLCA is not None and rightLCA is not None: 
            return root
        if leftLCA is not None: 
            return leftLCA
        else: 
            return rightLCA


if __name__ =='__main__':
    tree = BinaryTree(4)
    tree.put(tree.root,2)
    tree.put(tree.root,3)
    tree.put(tree.root,1)
    tree.put(tree.root,6)
    tree.put(tree.root,5)
    tree.put(tree.root,7)
    lcaFound = tree.findLCA(tree.root, 3,7)
    if lcaFound is not None and tree.get(tree.root,3) and tree.get(tree.root,7):
        print("LCA of key:3 and key:7  = " + str(lcaFound.key))
    else:
        print("LCA of key:3 and key:7 does not exist")
    lcaFound = tree.findLCA(tree.root, 4, 10)
    if lcaFound is not None and tree.get(tree.root,4) and tree.get(tree.root,10):
        print("LCA of key:4 and key:10 is" + str(lcaFound.key))
    else:
        print("LCA of key:4 and key:10 does not exist")




